File: backend/src/main/java/com/example/Services/ProductRowImportService.java
package com.example.Services;

import com.example.Repository.AuthorRepository;
import com.example.Repository.GenereRepository;
import com.example.Repository.LanguageRepository;
import com.example.Repository.ProductRepository;
import com.example.Repository.ProductTypeRepository;
import com.example.Repository.PublisherRepository;
import com.example.models.Author;
import com.example.models.Genere;
import com.example.models.Language;
import com.example.models.Product;
import com.example.models.Product_Type;
import com.example.models.Publisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Service
public class ProductRowImportService {

   
    private static final Map<String, String> LANGUAGE_TRANSLATION = Map.of(
            "मराठी", "Marathi",
            "हिंदी", "Hindi",
            "कोंकणी", "Konkani",
            "english", "English"
    );

    private static final Map<String, String> PRODUCT_TYPE_TRANSLATION = Map.of(
            "e-book", "eBook",
            "e-audio books", "Audiobook",
            "e-comics", "Comics"
    );


    private static final BigDecimal RENT_PRICE_PERCENT_OF_SALE = new BigDecimal("0.10");
    private static final int DEFAULT_MIN_RENT_DAYS = 1;


    private static final long OFFER_VALID_YEARS = 10;

    private final ProductRepository productRepository;
    private final AuthorRepository authorRepository;
    private final PublisherRepository publisherRepository;
    private final GenereRepository genereRepository;
    private final LanguageRepository languageRepository;
    private final ProductTypeRepository productTypeRepository;

    public ProductRowImportService(ProductRepository productRepository,
                                   AuthorRepository authorRepository,
                                   PublisherRepository publisherRepository,
                                   GenereRepository genereRepository,
                                   LanguageRepository languageRepository,
                                   ProductTypeRepository productTypeRepository) {
        this.productRepository = productRepository;
        this.authorRepository = authorRepository;
        this.publisherRepository = publisherRepository;
        this.genereRepository = genereRepository;
        this.languageRepository = languageRepository;
        this.productTypeRepository = productTypeRepository;
    }

    public enum Outcome {
        CREATED, SKIPPED
    }


    public record RowResult(Outcome outcome, String message, Integer productId) {
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public RowResult importRow(ExcelProductImportService.ParsedProductRow row) {

        
        boolean isPackageRow = "Package".equalsIgnoreCase(row.attributeSet())
                || "Yes".equalsIgnoreCase(row.isPackage());
        if (isPackageRow) {
            return new RowResult(Outcome.SKIPPED,
                    "Skipped - '" + row.prodName() + "' is a library package, not a book. "
                            + "Add packages from Admin > Products instead.",
                    null);
        }

        Optional<Product> existingProduct = productRepository.findByProductNameIgnoreCase(row.prodName());
        if (existingProduct.isPresent()) {
           
            return new RowResult(Outcome.SKIPPED,
                    "Skipped - '" + row.prodName() + "' is already in the catalogue.",
                    existingProduct.get().getProductId());
        }

        if (row.price() == null) {
            throw new IllegalArgumentException("No price given for '" + row.prodName() + "'");
        }

        String[] segments = splitTypeLanguageCategory(row.typeLanguageCategory());

        Product product = new Product();
        product.setProductName(truncate(row.prodName(), 150));
        product.setProductNameEnglish(truncate(row.nameEnglish(), 150));
        product.setProductType(resolveProductType(segments.length > 0 ? segments[0] : "eBook"));
        product.setLanguage(resolveLanguage(segments.length > 1 ? segments[1] : "English"));
        product.setGenere(resolveGenere(segments.length > 2 ? segments[2] : "General"));

        if (row.author() != null && !"package".equalsIgnoreCase(row.author())) {
            product.setAuthor(resolveAuthor(row.author()));
        }
        if (row.publisher() != null) {
            product.setPublisher(resolvePublisher(row.publisher()));
        }

        product.setProductBaseprice(row.price());
        applySpecialPrice(product, row);

        product.setProductDescriptionLong(stripHtml(row.description()));
        product.setProductDescriptionShort(truncate(stripHtml(row.shortDescription()), 255));

       
        product.setProductIsbn("BU-" + UUID.randomUUID().toString().replace("-", "").substring(0, 12).toUpperCase());

        applyAvailability(product, row.availability());

    

        product = productRepository.save(product);

        return new RowResult(Outcome.CREATED,
                "Created product #" + product.getProductId() + " '" + row.prodName() + "'",
                product.getProductId());
    }

    private String[] splitTypeLanguageCategory(String raw) {
        if (raw == null || raw.isBlank()) {
            return new String[0];
        }
        String[] parts = raw.split("/");
        for (int i = 0; i < parts.length; i++) {
            parts[i] = parts[i].trim();
        }
        return parts;
    }

    private Product_Type resolveProductType(String rawType) {
        String translated = PRODUCT_TYPE_TRANSLATION.getOrDefault(rawType.trim().toLowerCase(), rawType.trim());
        return productTypeRepository.findByTypeDescIgnoreCase(translated)
                .orElseGet(() -> {
                    Product_Type created = new Product_Type();
                    created.setTypeDesc(truncate(translated, 50));
                    return productTypeRepository.save(created);
                });
    }

    private Language resolveLanguage(String rawLanguage) {
        String translated = LANGUAGE_TRANSLATION.getOrDefault(rawLanguage.trim().toLowerCase(), rawLanguage.trim());
        return languageRepository.findByLanguageDescIgnoreCase(translated)
                .orElseGet(() -> {
                    Language created = new Language();
                    created.setLanguageDesc(truncate(translated, 50));
                    return languageRepository.save(created);
                });
    }

    private Genere resolveGenere(String rawGenre) {
        return genereRepository.findByGenereDescIgnoreCase(rawGenre)
                .orElseGet(() -> {
                    Genere created = new Genere();
                    created.setGenereDesc(truncate(rawGenre, 50));
                    return genereRepository.save(created);
                });
    }

    private Author resolveAuthor(String rawName) {
        String name = rawName.trim();
        return authorRepository.findByNameIgnoreCase(name)
                .orElseGet(() -> {
                    Author created = new Author();
                    created.setName(truncate(name, 100));
                    return authorRepository.save(created);
                });
    }

   
    private Publisher resolvePublisher(String rawName) {
        String name = rawName.trim();
        return publisherRepository.findByNameIgnoreCase(name)
                .orElseGet(() -> {
                    Publisher created = new Publisher();
                    created.setName(truncate(name, 100));
                    created.setEmail("bulk-upload+" + UUID.randomUUID() + "@placeholder.local");
                    return publisherRepository.save(created);
                });
    }

    private void applySpecialPrice(Product product, ExcelProductImportService.ParsedProductRow row) {
        if (row.specialPrice() == null || row.specialPrice().compareTo(row.price()) >= 0) {
            return; // no special price, or it isn't actually cheaper - leave the base price as-is
        }
        product.setProductOfferprice(row.specialPrice());
        product.setProductOffPriceExpirydate(LocalDate.now().plusYears(OFFER_VALID_YEARS));
    }

    
    private void applyAvailability(Product product, String availability) {
        String text = availability == null ? "" : availability.toLowerCase();

        boolean rentable = text.contains("rent");
        boolean lendable = text.contains("package");

        product.setRentable(rentable);
        product.setLibrary(lendable);

        if (rentable) {
            BigDecimal basis = product.getProductOfferprice() != null
                    ? product.getProductOfferprice()
                    : product.getProductBaseprice();
            product.setRentPerDay(basis.multiply(RENT_PRICE_PERCENT_OF_SALE).setScale(2, RoundingMode.HALF_UP));
            product.setMinRentDays(DEFAULT_MIN_RENT_DAYS);
        }
    }

    
    private String truncate(String value, int maxLength) {
        if (value == null || value.length() <= maxLength) {
            return value;
        }
        return value.substring(0, maxLength);
    }

    private String stripHtml(String html) {
        if (html == null) {
            return null;
        }
        String withBreaks = html
                .replaceAll("(?i)<br\\s*/?>", "\n")
                .replaceAll("(?i)</p>", "\n\n")
                .replaceAll("(?i)<[^>]+>", "");
        String unescaped = withBreaks
                .replace("&nbsp;", " ")
                .replace("&ldquo;", "“")
                .replace("&rdquo;", "”")
                .replace("&lsquo;", "‘")
                .replace("&rsquo;", "’")
                .replace("&mdash;", "—")
                .replace("&ndash;", "–")
                .replace("&hellip;", "…")
                .replace("&quot;", "\"")
                .replace("&#39;", "'")
                .replace("&amp;", "&")
                .replace("&lt;", "<")
                .replace("&gt;", ">");
        return unescaped.replaceAll("[ \\t]+", " ").replaceAll("\\n{3,}", "\n\n").trim();
    }
}

File: dotnet-backend/BookwormApi/Service/ProductRowImportService.cs
using System.Text.RegularExpressions;
using BookwormApi.Data;
using BookwormApi.Models;
using Microsoft.EntityFrameworkCore;

namespace BookwormApi.Service;

public class ProductRowImportService : IProductRowImportService
{
    private const decimal RentPricePercentOfSale = 0.10m;
    private const int DefaultMinRentDays = 1;
    private const int OfferValidYears = 10;

    private static readonly Dictionary<string, string> LanguageTranslation = new()
    {
        ["मराठी"] = "Marathi",
        ["हिंदी"] = "Hindi",
        ["कोंकणी"] = "Konkani",
        ["english"] = "English"
    };

    private static readonly Dictionary<string, string> ProductTypeTranslation = new()
    {
        ["e-book"] = "eBook",
        ["e-audio books"] = "Audiobook",
        ["e-comics"] = "Comics"
    };

    private readonly BookwormDbContext _db;

    public ProductRowImportService(BookwormDbContext db)
    {
        _db = db;
    }

    public async Task<RowImportResult> ImportRowAsync(ParsedProductRow row)
    {
        var prodName = row.ProdName ?? string.Empty;

        var isPackageRow = string.Equals(row.AttributeSet, "Package", StringComparison.OrdinalIgnoreCase) ||
                            string.Equals(row.IsPackage, "Yes", StringComparison.OrdinalIgnoreCase);
        if (isPackageRow)
        {
            return new RowImportResult(RowImportStatus.Skipped,
                $"Skipped - '{prodName}' is a library package, not a book. Add packages from Admin > Products instead.", null);
        }

        var existing = await _db.Products.FirstOrDefaultAsync(p => p.ProductName.ToLower() == prodName.ToLower());
        if (existing is not null)
        {
            return new RowImportResult(RowImportStatus.Skipped, $"Skipped - '{prodName}' is already in the catalogue.", existing.ProductId);
        }

        if (row.Price is null)
        {
            throw new ArgumentException($"No price given for '{prodName}'");
        }

        var segments = SplitTypeLanguageCategory(row.TypeLanguageCategory);

        var product = new Product
        {
            ProductName = Truncate(prodName, 150) ?? string.Empty,
            ProductNameEnglish = Truncate(row.NameEnglish, 150),
            ProductBaseprice = row.Price.Value
        };

        product.ProductTypeId = (await ResolveProductTypeAsync(segments.Length > 0 ? segments[0] : "eBook")).TypeId;
        product.LanguageId = (await ResolveLanguageAsync(segments.Length > 1 ? segments[1] : "English")).LanguageId;
        product.GenereId = (await ResolveGenereAsync(segments.Length > 2 ? segments[2] : "General")).GenereId;

        if (row.Author is not null && !string.Equals(row.Author, "package", StringComparison.OrdinalIgnoreCase))
        {
            product.AuthorId = (await ResolveAuthorAsync(row.Author)).AuthorId;
        }
        if (row.Publisher is not null)
        {
            product.PublisherId = (await ResolvePublisherAsync(row.Publisher)).PublisherId;
        }

        ApplySpecialPrice(product, row);

        product.ProductDescriptionLong = StripHtml(row.Description);
        product.ProductDescriptionShort = Truncate(StripHtml(row.ShortDescription), 255);
        product.ProductIsbn = $"BU-{Guid.NewGuid():N}"[..15].ToUpperInvariant();

        ApplyAvailability(product, row.Availability);

        _db.Products.Add(product);
        await _db.SaveChangesAsync();

        return new RowImportResult(RowImportStatus.Created, $"Created product #{product.ProductId} '{prodName}'", product.ProductId);
    }

    private static string[] SplitTypeLanguageCategory(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return Array.Empty<string>();
        return raw.Split('/').Select(s => s.Trim()).ToArray();
    }

    private async Task<ProductType> ResolveProductTypeAsync(string raw)
    {
        var translated = ProductTypeTranslation.GetValueOrDefault(raw.Trim().ToLower(), raw.Trim());
        var existing = await _db.ProductTypes.FirstOrDefaultAsync(t => t.TypeDesc.ToLower() == translated.ToLower());
        if (existing is not null) return existing;

        var created = new ProductType { TypeDesc = Truncate(translated, 50) ?? translated };
        _db.ProductTypes.Add(created);
        await _db.SaveChangesAsync();
        return created;
    }

    private async Task<Language> ResolveLanguageAsync(string raw)
    {
        var translated = LanguageTranslation.GetValueOrDefault(raw.Trim().ToLower(), raw.Trim());
        var existing = await _db.Languages.FirstOrDefaultAsync(l => l.LanguageDesc.ToLower() == translated.ToLower());
        if (existing is not null) return existing;

        var created = new Language { LanguageDesc = Truncate(translated, 50) ?? translated };
        _db.Languages.Add(created);
        await _db.SaveChangesAsync();
        return created;
    }

    private async Task<Genere> ResolveGenereAsync(string raw)
    {
        var trimmed = raw.Trim();
        var existing = await _db.Generes.FirstOrDefaultAsync(g => g.GenereDesc.ToLower() == trimmed.ToLower());
        if (existing is not null) return existing;

        var created = new Genere { GenereDesc = Truncate(trimmed, 50) ?? trimmed };
        _db.Generes.Add(created);
        await _db.SaveChangesAsync();
        return created;
    }

    private async Task<Author> ResolveAuthorAsync(string raw)
    {
        var trimmed = raw.Trim();
        var existing = await _db.Authors.FirstOrDefaultAsync(a => a.Name.ToLower() == trimmed.ToLower());
        if (existing is not null) return existing;

        var created = new Author { Name = Truncate(trimmed, 100) ?? trimmed };
        _db.Authors.Add(created);
        await _db.SaveChangesAsync();
        return created;
    }

    private async Task<Publisher> ResolvePublisherAsync(string raw)
    {
        var trimmed = raw.Trim();
        var existing = await _db.Publishers.FirstOrDefaultAsync(p => p.Name.ToLower() == trimmed.ToLower());
        if (existing is not null) return existing;

        var created = new Publisher
        {
            Name = Truncate(trimmed, 100) ?? trimmed,
            Email = $"bulk-upload+{Guid.NewGuid()}@placeholder.local"
        };
        _db.Publishers.Add(created);
        await _db.SaveChangesAsync();
        return created;
    }

    private static void ApplySpecialPrice(Product product, ParsedProductRow row)
    {
        if (row.SpecialPrice is null || row.SpecialPrice >= row.Price)
        {
            return;
        }
        product.ProductOfferprice = row.SpecialPrice;
        product.ProductOffPriceExpirydate = DateOnly.FromDateTime(DateTime.Today).AddYears(OfferValidYears);
    }

    private static void ApplyAvailability(Product product, string? availability)
    {
        var text = (availability ?? string.Empty).ToLower();
        product.Rentable = text.Contains("rent");
        product.Library = text.Contains("package");

        if (product.Rentable)
        {
            var basis = product.ProductOfferprice ?? product.ProductBaseprice;
            product.RentPerDay = Math.Round(basis * RentPricePercentOfSale, 2, MidpointRounding.AwayFromZero);
            product.MinRentDays = DefaultMinRentDays;
        }
    }

    private static readonly Regex BrTagRegex = new("<br\\s*/?>", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex AnyTagRegex = new("<[^>]*>", RegexOptions.Compiled);
    private static readonly Regex SpaceRunRegex = new("[ \t]+", RegexOptions.Compiled);
    private static readonly Regex NewlineRunRegex = new("\n{3,}", RegexOptions.Compiled);

    internal static string? StripHtml(string? html)
    {
        if (string.IsNullOrEmpty(html)) return html;

        var text = BrTagRegex.Replace(html, "\n");
        text = text.Replace("</p>", "\n\n", StringComparison.OrdinalIgnoreCase);
        text = AnyTagRegex.Replace(text, "");

        text = text.Replace("&nbsp;", " ")
            .Replace("&ldquo;", "“")
            .Replace("&rdquo;", "”")
            .Replace("&lsquo;", "‘")
            .Replace("&rsquo;", "’")
            .Replace("&mdash;", "—")
            .Replace("&ndash;", "–")
            .Replace("&hellip;", "…")
            .Replace("&quot;", "\"")
            .Replace("&#39;", "'")
            .Replace("&amp;", "&")
            .Replace("&lt;", "<")
            .Replace("&gt;", ">");

        text = SpaceRunRegex.Replace(text, " ");
        text = NewlineRunRegex.Replace(text, "\n\n");

        return text.Trim();
    }

    internal static string? Truncate(string? value, int maxLength)
    {
        if (value is null) return null;
        return value.Length > maxLength ? value[..maxLength] : value;
    }
}
